### DISCUSSION TITLE
Is there a way to bind different actions when the cursor is at the beginning of the line and another otherwise?

### DISCUSSION THREAD

**Discussion by yerlaser** [nushell/nushell#14200] [Q&A]
I would like to bind `space` to open history menu when pressed at the start of the line and insert space normally otherwise.

**Comment by MatrixManAtYrService** [comment]
It feels possible, but I can only offer a few thoughts in that direction. Hopefully somebody who knows better comes along and shows the way more concisely.

In my config I do something similar where Ctrl+T does something different based on a trailing space:

- `rm foo ` # use fzf to start searching for a second file to delete
- `rm foo` # use fzf to start searching for files to delete which start with foo

Here's the code that does that: https://github.com/MatrixManAtYrService/configs/blob/f9c19d4663ac25a1b52a7d3948aea1a582c61094/home-manager/config/nushell/config.nu#L793

You can see the valid keycodes and modifiers like this:
```
keybindings list --keycodes
keybindings list --modifiers
```

One way would be to use
```nushell
modifier: none
keycode: space
event: [
    {
        send: ExecuteHostCommand
        cmd: "commandline edit ( ...
```

And then you'd write something which evaluates to a space if the output of `commandline` is not empty, or open the history menu if it is.

In my case, I pipe the output of `history` into `fzf` ([here](https://github.com/MatrixManAtYrService/configs/blob/f9c19d4663ac25a1b52a7d3948aea1a582c61094/home-manager/config/nushell/config.nu#L781)) which lets me handle everything in `ExecuteHostCommand`.  If you didn't want to use something like `fzf` you'd need to figure out how to launch the default search history. The default config entry for it looks like this:

```nushell
        {
            name: search_history
            modifier: control
            keycode: char_q
            mode: [emacs, vi_normal, vi_insert]
            event: { send: searchhistory }
        }
```

With `searchhistory` being a type of event and not a standalone command.  I'm not quite sure how you'd invoke that from inside ExecuteHostCommand.  

An alternative approach would be to avoid summoning the history picker from inside `ExecuteHostCommand` and instead work with `until`, sort of like how this keybind does it:

```nushell
        {
            name: move_right_or_take_history_hint
            modifier: control
            keycode: char_f
            mode: emacs
            event: {
                until: [
                    { send: historyhintcomplete }
                    { send: menuright }
                    { send: right }
                ]
            }
        }
```

But I'm not sure what it takes to register custom events.  I think you'd want something like this:

```
            event: {
                until: [
                    { send: append-space-if-commandline-is-not-empty-fail-otherwise } 
                    { send: searchhistory }
                ]
            }
```

**Reply by yerlaser** [reply]
Hi,

Thanks a lot for the very detailed answer!

I will check your solution for detection of start line (or empty line).

You are right, though, I won't be able to open the history menu this way.

I hope that one day it will be possible to send a command like *move left* and if it fails use the built-in **until mechanism** to launch the menu, otherwise move back and add the space.

As per the documentation currently all these commands succeed even if there's nowhere to move to the left.

**Comment by Lockszmith-GH** [comment]
Take a look at [atuin](https://atuin.sh), maybe you can use that instead of the internal history picker.

**Reply by MatrixManAtYrService** [reply]
I've considered switching from fzf to atuin, it looks pretty cool.  Although from a get-to-know-nushell perspective I'm still curious of one can write custom events or access existing events as commands such that this could be done without 3rd party software.  I could be overlooking something but they don't seem like worlds that need to be as separate as they feel.

**Reply by yerlaser** [reply]
> Take a look at [atuin](https://atuin.sh), maybe you can use that instead of the internal history picker.

Thanks for the suggestion, but, no, no and no :)

There's no way I'll let my command history be dumped into the Internet. More so, to allow it to come back into another device after being in the Internet.

**Reply by Lockszmith-GH** [reply]
Don't send it to the internet, I'm not doing it either (although it's completely encrypted), it's all local for me.

The only reason I'm suggesting it, as an alternative to the history invocation, which seems to be the action that can't be invoked.

**Reply by yerlaser** [reply]
OK. I see.
Thanks for the suggestion, I'll look into it.
However, nu's history menu perfectly satisfies me, I just need to find a way to invoke it with a very easy shortcut.
Backspace and Enter at the beginning of the line are useless so they are good candidates as well.
